File: server/internal/dto/moment.go
/*
项目名称：JeriBlog
文件名称：moment.go
创建时间：2026-04-16 15:00:50

系统用户：Jerion
作　　者：Jerion
联系邮箱：416685476@qq.com
功能描述：动态数据传输对象
*/

package dto

import "jeri_blog/pkg/utils"

// ============ 前台动态请求 ============

// ListMomentsForWebRequest 前台动态列表请求
type ListMomentsForWebRequest struct {
	Page     int `form:"page" binding:"omitempty,min=1"`
	PageSize int `form:"page_size" binding:"omitempty,min=1,max=1000"`
}

// FetchLinkMetadataRequest 获取链接元数据请求
type FetchLinkMetadataRequest struct {
	URL string `json:"url" binding:"required,url"`
}

// LinkMetadataResponse 链接元数据响应
type LinkMetadataResponse struct {
	Title   string `json:"title"`
	Favicon string `json:"favicon"`
}

// ParseVideoRequest 解析视频请求
type ParseVideoRequest struct {
	URL string `json:"url" binding:"required"`
}

// ParseVideoResponse 解析视频响应
type ParseVideoResponse struct {
	Platform string `json:"platform"` // 'bilibili' | 'youtube'
	VideoID  string `json:"video_id"` // 视频ID
}

// ============ 通用动态响应 ============

// MomentContent 动态内容结构
type MomentContent struct {
	Text     string         `json:"text,omitempty"`     // 文本
	Images   []string       `json:"images,omitempty"`   // 图片列表
	Location string         `json:"location,omitempty"` // 位置信息
	Tags     string         `json:"tags,omitempty"`     // 标签
	Link     *MomentLink    `json:"link,omitempty"`     // 外链
	Music    *MomentMusic   `json:"music,omitempty"`    // 音乐（基于MetingJS）
	Video    *MomentVideo   `json:"video,omitempty"`    // 视频（本地或在线）
	Audio    *MomentAudio   `json:"audio,omitempty"`    // 音频（本地或在线）
	Book     map[string]any `json:"book,omitempty"`     // 书籍
	Movie    map[string]any `json:"movie,omitempty"`    // 电影
}

// MomentLink 外链结构
type MomentLink struct {
	URL     string `json:"url"`
	Title   string `json:"title"`
	Favicon string `json:"favicon"`
}

// MomentMusic 音乐结构（基于 MetingJS）
type MomentMusic struct {
	Server string `json:"server"` // 音乐平台：netease, tencent
	Type   string `json:"type"`   // 类型：song, playlist, album, artist
	ID     string `json:"id"`     // 音乐ID
}

// MomentVideo 视频结构
type MomentVideo struct {
	URL      string `json:"url"`                // 视频URL（本地视频或在线视频链接）
	Platform string `json:"platform,omitempty"` // 平台：bilibili, youtube（本地视频为空）
	VideoID  string `json:"video_id,omitempty"` // 视频ID（在线视频的ID，本地视频为空）
}

// MomentAudio 音频结构
type MomentAudio struct {
	URL string `json:"url"` // 音频URL（本地音频或在线音频链接）
}

// MomentForWebResponse 前台动态响应
type MomentForWebResponse struct {
	ID          uint            `json:"id"`
	Content     MomentContent   `json:"content"`
	IsPublish   bool            `json:"is_publish"`
	PublishTime *utils.JSONTime `json:"publish_time"`
}

// ============ 后台动态管理请求 ============

// ListMomentRequest 后台动态列表请求（支持筛选）
type ListMomentRequest struct {
	Page      int      `form:"page" binding:"omitempty,min=1"`
	PageSize  int      `form:"page_size" binding:"omitempty,min=1,max=1000"`
	Keyword   string   `form:"keyword"`    // 搜索关键词（文本内容）
	Tags      []string `form:"tags"`       // 标签（多选）
	Location  string   `form:"location"`   // 发布地点
	IsPublish *bool    `form:"is_publish"` // 是否发布
	HasImages *bool    `form:"has_images"` // 是否有图片
	HasVideo  *bool    `form:"has_video"`  // 是否有视频
	HasAudio  *bool    `form:"has_audio"`  // 是否有音频
	HasMusic  *bool    `form:"has_music"`  // 是否有音乐
	HasLink   *bool    `form:"has_link"`   // 是否有链接
	StartTime string   `form:"start_time"` // 发布开始时间
	EndTime   string   `form:"end_time"`   // 发布结束时间
}

// CreateMomentRequest 创建动态请求
type CreateMomentRequest struct {
	Content     MomentContent   `json:"content" binding:"required"`
	IsPublish   bool            `json:"is_publish"`   // 是否发布，默认true
	PublishTime *utils.JSONTime `json:"publish_time"` // 发布时间（可选，不填则使用创建时间）
}

// UpdateMomentRequest 更新动态请求
type UpdateMomentRequest struct {
	Content     MomentContent   `json:"content" binding:"required"`
	IsPublish   bool            `json:"is_publish"`   // 是否发布
	PublishTime *utils.JSONTime `json:"publish_time"` // 发布时间（可编辑）
}

// ============ 后台动态管理响应 ============

// MomentListResponse 后台动态列表响应
type MomentListResponse struct {
	ID          uint            `json:"id"`
	Content     MomentContent   `json:"content"`
	IsPublish   bool            `json:"is_publish"`
	PublishTime *utils.JSONTime `json:"publish_time"`
}

File: admin/src/types/moment.ts
/*
项目名称：JeriBlog
文件名称：moment.ts
创建时间：2026-04-16 15:08:10

系统用户：Jerion
作　　者：Jerion
联系邮箱：416685476@qq.com
功能描述：类型定义 - moment类型
*/

// 动态内容类型
export interface MomentContent {
  text?: string;
  tags?: string;
  images?: string[];
  video?: {
    url: string;
    platform?: string;
    video_id?: string;
  };
  audio?: {
    url: string;
  };
  music?: {
    server: string;
    type: string;
    id: string;
  };
  link?: {
    url: string;
    title?: string;
    favicon?: string;
  };
  location?: string;
  book?: Record<string, unknown>;
  movie?: Record<string, unknown>;
}

// 动态实体
export interface Moment {
  id: number;
  content: MomentContent;
  is_publish: boolean; // 是否发布
  publish_time: string;
  created_at: string;
  updated_at: string;
}

// 创建动态请求
export interface CreateMomentRequest {
  content: MomentContent;
  publish_time?: string;
  is_publish?: boolean;
}

// 更新动态请求
export interface UpdateMomentRequest {
  content?: MomentContent;
  publish_time?: string;
  is_publish?: boolean;
}

// 获取链接信息请求
export interface FetchLinkRequest {
  url: string;
}

// 链接信息响应
export interface LinkInfo {
  title: string;
  favicon: string;
}

// 解析视频请求
export interface ParseVideoRequest {
  url: string;
}

// 视频信息响应
export interface VideoInfo {
  platform: string;
  video_id: string;
}

// 动态列表查询参数
export interface MomentListQuery {
  page: number;
  page_size: number;
  keyword?: string; // 搜索关键词（文本内容）
  tags?: string[]; // 标签
  location?: string; // 发布地点
  is_publish?: boolean; // 是否发布
  has_images?: boolean; // 是否有图片
  has_video?: boolean; // 是否有视频
  has_audio?: boolean; // 是否有音频
  has_music?: boolean; // 是否有音乐
  has_link?: boolean; // 是否有链接
  start_time?: string; // 发布开始时间
  end_time?: string; // 发布结束时间
}

// 分页数据
export interface MomentListData {
  list: Moment[];
  total: number;
  page: number;
  page_size: number;
}
